Add documentation

christiansakai opened the pull request:
@donkeybonks Sanity check for the first PR here, I added two lines of documentation on `Payload.hs`

* Does this seem okay?
* Looking at the Contributing Guidelines, I have a few questions:
   * What do we use for linting?
   * I'm not sure about the `cabal` infrastructure. I am using `stack` to build, test, and generate the docs.

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -64,7 +64,7 @@ instance FromJSON OwnerType where
           "organization"  -> pure OwnerOrganization
           _               -> fail $ "Unknown owner type: " ++ T.unpack t
 
-
```
Please preserve the white space spacing here, create new white space for documentation rather than inserting into existing space. i.e. there should be no red lines in this diff

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -64,7 +64,7 @@ instance FromJSON OwnerType where
           _               -> fail $ "Unknown owner type: " ++ T.unpack t
 
-
+-- | Github URL
```
Documentation should be valid English sentences that end with a full stop, for example:
```
-- | Triggered when an issue comment is created, edited, or deleted.
-- See <https://developer.github.com/v3/activity/events/types/#issuecommentevent>.
```
Note also that 'Github' should not be used in place of 'GitHub', ie see the copyright at the bottom of this page, '© 2018 GitHub, Inc.'

In this instance
```
-- | Represents a URL that would be suitable to query for more information. The GitHub API only returns valid URLs.
```

christiansakai replied:
What is the significance of the statement `The GitHub API only returns valid 'URL's`

donkeybonks replied:
Good question.

GitHub only returns URLs that are valid, ie they start with https:// or such. If you can imagine a ‘validateUrl :: Text -> Maybe URL’ . That’s the information that is stored by the opaque type URL.

getUrl discards this information.

christiansakai replied:
I am not sure where do I see `getUrl` discard an invalid URL? Or maybe I misread your statement? As far as the notation goes, it is only unboxing `URL Text` to `Text`.

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -73,6 +73,7 @@ instance NFData URL where rnf = genericRnf
 instance FromJSON URL where
   parseJSON = withText "URL" (pure . URL)
 
+-- | Convert Github URL to Text
```
You can link types here. When you generate the documentation they will form hyperlinks. For example
```
-- | Converts a 'URL' to 'Text'.
```

christiansakai replied:
I agree with you.

donkeybonks commented:
1. hlint ( https://github.com/ndmitchell/hlint ) is run in Travis automatically for linting
2. We don't use the cabal infrastructure, stack handles everything so there is no need to worry. For most operations stack ~ cabal operate the same, except stack just scopes the compiler and packages properly to the resolvers specified in `stack.yaml`. The cabal file is automatically generated by hpack ( https://github.com/sol/hpack ) from `package.yaml`.

donkeybonks commented:
Hey, @christiansakai any issues?

christiansakai commented:
Hey @donkeybonks no issues at all, sorry I've been concentrating on midterms (this week and next week) is my midterm week. I'll get back ASAP. If you want to push something, please do so.

christiansakai commented:
I notice everytime I run `stack haddock` it proceeds to build the entire library. e.g, it builds the `.stack-work/dist/` folder. Is this the default behavior?

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -107,7 +111,7 @@ data HookIssue = HookIssue
 instance NFData HookIssue where rnf = genericRnf
 
-
+-- | Represents the "repository" field in all type of payload.
```
`type` -> `types`

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -199,7 +203,7 @@ data HookRepositoryLabel = HookRepositoryLabel
 instance NFData HookRepositoryLabel where rnf = genericRnf
 
-
+-- | Represents the "user" field in all type of payload.
```
`type` -> `types`.

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -233,7 +237,7 @@ data HookSimpleUser = HookSimpleUser
 instance NFData HookSimpleUser where rnf = genericRnf
 
-
+-- | Represents the "organization" field in all type of payload.
```
`type` -> `types`

donkeybonks commented on src/GitHub/Data/Webhooks/Payload.hs:
```diff
@@ -377,7 +385,8 @@ data HookProjectColumn = HookProjectColumn
 
-
+-- | Represents the "issue.labels" field in the
+-- 'IssueCommentEvent' and 'IssueEvent' payload.
```
`payload` -> `payloads`.

donkeybonks commented:
It needs to build the entire library because it needs the type information to generate the docs and that requires it to generate the interface files (which in turn need Haskell bytecode for inline functions), or something like that.

christiansakai commented:
@donkeybonks 

Yes, I will squash the commits once it is done. There are still other sections to be filled up. I have some questions about those:
1. Should I add documentation about the `FIXME` items on `Payload.hs`?
2. How should the Module Headers documentation look like?
3. For `Events.hs`, can you provide an example of what to write for `EventHasSender` and `EventHasRepo`?

donkeybonks commented:
@christiansakai 

1. Please do not put FIXMEs in documentation, rather just leave them as regular comments.
2. See here https://github.com/onrock-eng/github-webhooks/issues/5#issuecomment-372197017 .
3. `EventHasSender` and `EventHasRepo` are for events with a "sender" and "repository" field respectively. This is standard across most events, such that i.e. if you have a list `forall a. EventHasRepo a => a` you can get the repo for each regardless of the actual type. A simple non-technical description is suffice.

donkeybonks commented:
Hey, FYI there will probably be another minor release within the next 2-3 weeks for GHC-7. Exact roadmap will be published soon.

christiansakai commented:
@donkeybonks What is the significance of that? Just wondering.

Also, I've been busy again with school work, sorry about that. I'll try to put more updates to the PR.

donkeybonks commented:
Significance may just be a milestone; If the PR is merged before the next minor release it will be pushed to Hackage.

christiansakai commented:
@donkeybonks 

I noticed that in most of Haskell documentations, they don't add the short module description, but rather the long description, so I did that for `Payload.hs` and `Event.hs`. Please take a look and let me know if you want to change anything.

For `Payload.hs`, the completion rate is 85%, the missing ones are the `FIXME` ones:
- HookSimpleUser (src/GitHub/Data/Webhooks/Payload.hs:238)
- HookMembership (src/GitHub/Data/Webhooks/Payload.hs:333)
- HookCommit (src/GitHub/Data/Webhooks/Payload.hs:411)
- HookRelease (src/GitHub/Data/Webhooks/Payload.hs:425)
- HookPullRequest (src/GitHub/Data/Webhooks/Payload.hs:450)

I'm going to work on `Event.hs` now

christiansakai commented:
I'm currently trying to understand `Event.hs`.

I added documentation on `EventHasSender` and `EventHasRepo`, please let me know if you want to change it.

I look at `CommitCommentEventAction` and added documentation like this

> Represents the "action" field in the 'CommitCommentEvent' payload.

because I realized that this is the `action` field for https://developer.github.com/v3/activity/events/types/#commitcommentevent

I'm wondering why this is not in `Payload.hs`? So I wonder if the wording should be 

> Represents the "action" field in the 'CommitCommentEvent' event.

Or maybe I should change all my reference of `payload` to just `JSON payload` to make it clear that this is speaking about the JSON object and not about the code in `Payload.hs`.

donkeybonks commented:
Apologies in delay;

* Need to do some significant work to this repo soon
* So I think it would be a good idea to squash, rebase and merge at this point
* Will answer other questions soon but feel free to open another PR
* As you squash, you can also add yourself to AUTHORS

donkeybonks commented:
We believe it is better to move fast incrementally at this point than trying to maintain huge changesets. 80/20 rule.

christiansakai commented:
@donkeybonks 

Are Authors == Maintainers?

Currently Authors are located in three places:
`AUTHORS`
`package.yaml`

Maintainers are located in two places:
`package.yaml`
`Payload.hs`
`Event.hs`

Should I add me to these files?

donkeybonks commented:
Hi,

Authors /= maintainers, so whilst an AUTHOR may commit a piece of code, maintainers are slightly different as they take responsibility for releases and roadmap.

You can add yourself to AUTHORS and people looking for a domain expert here will be able to contact you.

christiansakai commented:
@donkeybonks I rebased and pushed. Is this okay?

donkeybonks commented:
You will need to run hpack to regenerate the cabal file with the updated authors field

christiansakai commented:
@donkeybonks Just generated the cabal file using `hpack`

donkeybonks commented:
Looks good!
